### PR TITLE
Gitc 617 - Url unfurl grants new

### PR DESCRIPTION
**Pull request by ksolo**
##### Description

Unfurling the `grants/new` page was showing GitHub. This was caused by the login required decorated that redirected requests to GitHub for signing in. 

This PR:
* Adds tests for the `/grants/new` GET route
* Adds tests for the `/grants/new` POST route
* Creates a `GrantSubmissionView` class to allow separation of login requirements to the GET and POST routes
* Adds conditional to the template to show the form or the request to log in via GitHub

https://user-images.githubusercontent.com/103829/148455371-0f8c725b-6db8-439c-bfac-ae9b502c6be4.mov

##### Refers/Fixes

GITC-617

##### Testing

Pytest for the view were added. Staging will verify the unfurl happens correctly.

### CODE REVIEW

File: app/git/tests/models/test_git_cache.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import pytest
+from faker import Faker
```
Comment (ksolo) [inline app/git/tests/models/test_git_cache.py:2]: Lint fix

File: app/git/tests/test_utils.py [hunk 2]
```diff
@@ -19,25 +19,22 @@
...
-    get_issue_comments, get_issue_timeline_events, github_connect, is_github_token_valid, org_name, patch_issue_comment,
-    post_issue_comment, post_issue_comment_reaction, repo_url, reset_token, revoke_token, _get_user, _get_repo, _get_issue,
-    _get_issue_comment
+    HEADERS, TOKEN_URL, _get_issue, _get_issue_comment, _get_repo, _get_user, build_auth_dict, delete_issue_comment,
+    get_github_emails, get_github_primary_email, get_issue_comments, get_issue_timeline_events, github_connect,
+    is_github_token_valid, org_name, patch_issue_comment, post_issue_comment, post_issue_comment_reaction, repo_url,
+    reset_token, revoke_token,
 )
-from github import NamedUser
+from github import Issue, IssueComment, NamedUser, Repository
```
Comment (ksolo) [inline app/git/tests/test_utils.py:37]: lint fix

File: app/grants/models/__init__.py [hunk 3]
```diff
@@ -32,7 +32,7 @@
 from .grant_stat import GrantStat
 from .grant_tag import GrantTag
 from .grant_type import GrantType
+from .hall_of_fame import GrantHallOfFame, GrantHallOfFameGrantee
```
Comment (ksolo) [inline app/grants/models/__init__.py:35]: lint fix

File: app/grants/templates/grants/_new.html [hunk 4]
```diff
@@ -39,418 +39,432 @@
...
-                <p class="font-bigger-1 font-weight-light mb-0">
-                  We're excited to learn about your project.
-                </p>
-                <p class="text-center">
-                  <button @click="showQuickStart()" type="button" class="font-weight-light font-bigger-1 btn btn-link">
-                    Open Grant Guide
-                  </button>
-                </p>
-              </div>
+      {% if user.is_authenticated %}
```
Comment (ksolo) [inline app/grants/templates/grants/_new.html:42]: The change to the template included a check for user authentication. The content in the if block is what was there before, no changes, the content in the else block is new and needs review

File: app/grants/templates/grants/_new.html [hunk 5]
```diff
@@ -39,418 +39,432 @@
...
                 </div>
               </div>
             </div>
-          </div>
 
-        </form>
-      </div>
+          </form>
+        </div>
+      {% else %}
```
Comment (ksolo) [inline app/grants/templates/grants/_new.html:455]: This is the beginning of the new content. Ends on ln467

File: app/grants/urls.py [hunk 6]
```diff
@@ -20,17 +20,18 @@
...
-    contribution_addr_from_grant_as_json, contribution_addr_from_grant_during_round_as_json,
-    contribution_addr_from_round_as_json, contribution_info_from_grant_during_round_as_json, create_matching_pledge_v1,
-    delete_collection, flag, get_clr_sybil_input, get_collection, get_collections_list, get_ethereum_cart_data,
-    get_grant_payload, get_grant_tags, get_grants, get_interrupted_contributions, get_replaced_tx, get_trust_bonus,
-    grant_activity, grant_details, grant_details_api, grant_details_contributions, grant_details_contributors,
-    grant_edit, grant_fund, grant_new, grants, grants_addr_as_json, grants_bulk_add, grants_by_grant_type,
-    grants_cart_view, grants_info, grants_landing, grants_type_redirect, hall_of_fame, ingest_contributions,
-    ingest_contributions_view, invoice, leaderboard, manage_ethereum_cart_data, new_matching_partner, profile,
-    quickstart, remove_grant_from_collection, save_collection, toggle_grant_favorite, upload_sybil_csv, verify_grant,
+    GrantSubmissionView, add_grant_from_collection, api_toggle_user_sybil, bulk_fund, bulk_grants_for_cart,
```
Comment (ksolo) [inline app/grants/urls.py:23]: the `GantSbumissionView` is new - the rest of the changes in this are linting fixes

File: app/grants/views.py [hunk 7]
```diff
@@ -3928,3 +3737,217 @@ def upload_sybil_csv(request):
         return JsonResponse({'success': 'failed'}, status=500)
 
     return JsonResponse({'success': 'ok'}, status=200)
+
+
+class GrantSubmissionView(View):
```
Comment (ksolo) [inline app/grants/views.py:3742]: This is the new class based view that replaced the previous function based view. In the previous view there was a top level conditional checking if the request is a POST, all of the code that was nested in that clause is now in the `post` method below. The code that was in the `else` of the clause is in the `get` method below.

**Review by Jer-Sch — APPROVED** [review]

File: app/grants/tests/views/test_grant_new_routes.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        client = Client(HTTP_USER_AGENT='chrome')
+        client.force_login(user)
+        response = client.post('/grants/new', grant_data)
+        response_data = json.loads(response.content)
+
+        assert response.status_code == 200
+        assert response_data.get('status') == 400
+        assert response_data.get('message') == 'error: payout_address is a mandatory parameter'
+
+    def test_zcash_addresses_must_be_tranparent_addresses(
```
[the remark this answers is not in the document]
Comment (ksolo) [reply app/grants/tests/views/test_grant_new_routes.py:137]: Good catch!

**Pull request merged**